### PR TITLE
Add gomoddirectives linter.

### PR DESCRIPTION
**Pull request by ldez**
https://github.com/ldez/gomoddirectives

```yml
  gomoddirectives:
    # Allow local `replace` directives. Default is false.
    replace-local: false
    # List of allowed `replace` directives. Default is empty.
    replace-allow-list:
      - launchpad.net/gocheck
    # Allow to not explain why the version has been retracted in the `retract` directives. Default is false.
    retract-allow-no-explanation: false
    # Forbid the use of the `exclude` directives. Default is false.
    exclude-forbidden: false
```

### CODE REVIEW

File: .golangci.example.yml
Comment (carnott-snap) [inline .golangci.example.yml:452]: I know my linter did not, so feel free to say we should add `gomodxxx`, but any chance we can expand the scope to restrict exclude and retract statements.

`retract` is new, and does generate stdout warnings when go-getting, but the code will build and run without them.

[`exclude`](https://github.com/golang/go/wiki/Modules#gomod) seems to be an escape hatch like replace, but for resolving conflicts or removing broken modules. it has a similar virality (must be in ALL main modules) to `replace`, so my gut say disallow.

Comment (ldez) [reply .golangci.example.yml:452]: yes the scope can be expanded. I need to get more information about `exclude` directives.

about `retract` directives I don't see what rule do you want to apply, could you explain?

File: .golangci.example.yml
Comment (carnott-snap) [inline .golangci.example.yml:454]: I am failing to think of a use case where I want to disallow all replace except for local. (please correct me if my understanding of how this works is wrong.) iirc we discussed the following use cases:
- disallow only local replace
- disallow all replace
- disallow all replace except for some modules

Comment (ldez) [reply .golangci.example.yml:454]: a use case: https://github.com/hashicorp/consul/blob/master/go.mod

File: .golangci.example.yml
Comment (ldez) [reply .golangci.example.yml:452]: the `exclude` are not transitive but I don't see how to handle that differently.

https://golang.org/ref/mod#go-mod-file-exclude

Comment (carnott-snap) [reply .golangci.example.yml:452]: for my repos, the lint for `exclude` and `replace` would be the same: disallow all. later we could add nuance to allow some `exclude` cases, but I have not seen it in the wild.

`retract` could be tricky. the ask is to "disallow any dependencies that themselves are retracted". now that I lay this out, it may be decent to do this as a separate linter, since the scope of "recurse through all dependencies and detect if their version is retracted" could get complex. your call. we may also run up against the fact that this resolution information only exists in `cmd/go/internal`.

another ask for `retract` is more a syntax thing: we should probably enforce that `modfile.File.Retract.Rationale != ""`.

File: .golangci.example.yml
Comment (carnott-snap) [reply .golangci.example.yml:454]: can you elaborate on how it is used or why it is required?

if this is a library it means that tests run from within and without would behave differently, something that seems bad. if this is a main module, it still seems gross to maintain, and I think using the `allow-list` syntax is cleaner.

File: .golangci.example.yml
Comment (carnott-snap) [inline .golangci.example.yml:457]: which of these does this allow:
- `replace launchpad.net/gocheck v0.0.0 => /path/to/gocheck`
- `replace launchpad.net/gocheck v0.0.0 => example.com/path/to/gocheck v0.0.0`
- `replace example.com/path/to/gocheck v0.0.0 => launchpad.net/gocheck v0.0.0`

Comment (ldez) [reply .golangci.example.yml:457]: currently only the last

File: .golangci.example.yml
Comment (ldez) [reply .golangci.example.yml:454]: I don't know exactly why they are using that, I saw that pattern several times.

I think it is a way of facilitating the development with the multi modules

Comment (carnott-snap) [reply .golangci.example.yml:454]: we can add it if people really want it, but that sounds like a things for local development, not something that should be merged or released

File: .golangci.example.yml
Comment (carnott-snap) [reply .golangci.example.yml:457]: that seems reasonable, as long as it is clear that `local: true` is not required to get the second

Comment (ldez) [reply .golangci.example.yml:457]: I change the behavior https://github.com/ldez/gomodreplace/blob/13fd25eefbe7e332e3d67b7497365363ae58e59f/gomodreplace_test.go#L29

File: .golangci.example.yml
Comment (carnott-snap) [inline .golangci.example.yml:456]: would nesting not be easier to maintain:
```suggestion
    replace:
      # Allow local `replace` directives. Default is false.
      local: false
      # List of allowed `replace` directives. Default is empty.
      allow-list:
```

File: .golangci.example.yml
Comment (carnott-snap) [inline .golangci.example.yml:461]: any reason to not use the same default as `replace`:
```suggestion
    # Allow the use of the `exclude` directives. Default is false.
    exclude: false
```

File: pkg/golinters/gomoddirectives.go
Comment (carnott-snap) [inline pkg/golinters/gomoddirectives.go:27]: if you use a custom type or line up the structs, you can do a direct assignment:
```suggestion
		opts = gomoddirectives.Options(settings)
```

File: .golangci.example.yml
Comment (ldez) [reply .golangci.example.yml:456]: I don't want to do that

File: .golangci.example.yml
Comment (ldez) [reply .golangci.example.yml:461]: because I don't want to exclude the use of `exclude` by default.

Comment (carnott-snap) [reply .golangci.example.yml:461]: then I would think we want to make `replace` off by default too?

File: pkg/golinters/gomoddirectives.go
Comment (ldez) [reply pkg/golinters/gomoddirectives.go:27]: it's not the same type, so it's not possible.

File: .golangci.example.yml
Comment (ldez) [reply .golangci.example.yml:461]: it's not the same goal ihmo, so I prefer is ban `replace` by default but allow `exclude` by default.

File: pkg/golinters/gomoddirectives.go
Comment (carnott-snap) [reply pkg/golinters/gomoddirectives.go:27]: actually, it is: https://play.golang.org/p/Gna-Uv1PAaS

File: .golangci.example.yml
Comment (carnott-snap) [reply .golangci.example.yml:456]: out of curiosity, why?

File: pkg/golinters/gomoddirectives.go
Comment (ldez) [reply pkg/golinters/gomoddirectives.go:27]:
```
# github.com/golangci/golangci-lint/pkg/golinters
pkg/golinters/gomoddirectives.go:24:33: cannot convert *settings (type "github.com/golangci/golangci-lint/pkg/config".GoModDirectivesSettings) to type gomoddirectives.Options
```

Comment (ldez) [reply pkg/golinters/gomoddirectives.go:27]: https://play.golang.org/p/0PbSOa7_aI4

**Review by carnott-snap — APPROVED** [review]
looks sane, meets the use cases we discussed; I have not tested this locally, but will trust the author/ci to assert its correctness. should we add some tests to verify anything, or are we trusting that is done in the upstream library?

**Comment by ldez** [comment]
@carnott-snap
> should we add some tests to verify anything,

It's not possible with the current test system, but I added a lot of tests in my lib.

> or are we trusting that is done in the upstream library?

FYI I'm a maintainer of golangci-lint, so yes we trust :smile:

File: pkg/golinters/gomoddirectives.go
Comment (carnott-snap) [reply pkg/golinters/gomoddirectives.go:27]: you changed the name of `ReplaceAllowLocal`: https://play.golang.org/p/o-97b9emRD_X

**Review by bombsimon — APPROVED** [review]
I trust you @ldez! ❤️

**Pull request merged**
